R/board_gcs.R

```diff
@@ -0,0 +1,262 @@
+#' several authentication options, as documented in its
+#' [main vignette](https://code.markedmondson.me/googleCloudStorageR/articles/googleCloudStorageR.html).
+#' The two main options are to create a service account key (a JSON file) or an
+#' authentication token; you can manage either using the [gargle](https://gargle.r-lib.org/) package.
+#'
+#' # Details
+#'
+#' * The functions in pins do not create a new bucket. You can create
+#'   a new bucket from R with [googleCloudStorageR::gcs_create_bucket()].
+#' * You can pass arguments for [googleCloudStorageR::gcs_upload] such as
+#'   `predefinedAcl` and `upload_type` through the dots of `pin_write()`.
+#' * `board_gcs()` is powered by the googleCloudStorageR package, which is a
+#'   suggested dependency of pins (not required for pins in general). If
+#'   you run into errors when deploying content to a server like
+#'   <https://shinyapps.io> or [Connect](https://posit.co/products/enterprise/connect/),
+#'   add `library(googleCloudStorageR)` to your app or document for [automatic
+#'   dependency discovery](https://support.posit.co/hc/en-us/articles/229998627-Why-does-my-app-work-locally-but-not-on-my-RStudio-Connect-server).
+#'
+#' @inheritParams new_board
+#' @param bucket Bucket name. You can only write to an existing bucket. Defaults
+#'   to [googleCloudStorageR::gcs_get_global_bucket()].
+#' @param prefix Prefix within this bucket that this board will occupy.
+#'   You can use this to maintain multiple independent pin boards within
+#'   a single GCS bucket. Will typically end with `/` to take advantage of
+#'   Google Cloud Storage's directory-like handling.
+#' @export
+#' @examples
+#' \dontrun{
+#' board <- board_gcs()
+#' board %>% pin_write(mtcars)
+#' board %>% pin_read("mtcars")
+#'
+#' # A prefix allows you to have multiple independent boards in the same pin.
+#' board_sales <- board_gcs("company-pins", prefix = "sales/")
+#' board_marketing <- board_gcs("company-pins", prefix = "marketing/")
+#' # You can make the hierarchy arbitrarily deep.
+#'
+#' # Pass arguments like `predefinedAcl` through the dots of `pin_write`:
+#' board %>% pin_write(mtcars, predefinedAcl = "publicRead")
+#' }
+board_gcs <- function(bucket = googleCloudStorageR::gcs_get_global_bucket(),
+                      prefix = NULL,
+                      versioned = TRUE,
+                      cache = NULL) {
+
+  check_installed("googleCloudStorageR")
+
+  # Check that have access to the bucket
+  invisible(googleCloudStorageR::gcs_get_bucket(bucket))
+
+  cache <- cache %||% board_cache_path(paste0("gcs-", bucket))
+  new_board_v1("pins_board_gcs",
+               name = "gcs",
+               bucket = bucket,
+               prefix = prefix,
+               cache = cache,
+               versioned = versioned
+  )
+}
+
+board_gcs_test <- function(...) {
+
+  skip_if_missing_envvars(
+    tests = "board_gcs()",
+    envvars = c("PINS_GCS_PASSWORD")
+  )
+
+  rlang::check_installed("sodium")
```

Comment:
FWIW we generally just assume suggested packages are available for testing: https://r-pkgs.org/dependencies-in-practice.html#sec-dependencies-in-suggests-in-tests